lib/identifiers/pubmed_id.rb

```diff
@@ -1,7 +1,40 @@
 module Identifiers
   class PubmedId
     def self.extract(str)
-      str.to_s.scan(/(?<=^|[[:space:]])0*(?!0)(\d+)(?=$|[[:space:]])/).flatten
+      new(str).extract
+    end
+
+    attr_reader :pubmed_candidate
+    private :pubmed_candidate
+
+    URL_REGEX = %r{^https?://(?:www\.)?ncbi.nlm.nih.gov/pubmed/}
+
+    def initialize(pubmed_candidate)
+      @pubmed_candidate = pubmed_candidate.to_s
+    end
+
+    def extract
+      extract_from_number | extract_from_pubmed_url
+    end
+
+    private
+
+    def extract_from_pubmed_url
+      return [] unless contains_pubmed_id?
+
+      Array(pubmed_id_from_url)
+    end
+
+    def contains_pubmed_id?
+      pubmed_candidate =~ URL_REGEX
+    end
+
+    def pubmed_id_from_url
+      pubmed_candidate.sub(URL_REGEX, '')
```

Comment:
Shall we also take into account any possible query/hash parameters? e.g. `https://www.ncbi.nlm.nih.gov/pubmed/6191871#cm6191871_69589`